loopy/preprocess.py

```diff
@@ -1999,8 +1999,142 @@ def realize_ilp(kernel):
+            elif expr.name in self.find_names:
+                self.result = True
+            return super(VariableFinder, self).map_variable(expr, args, kwargs)
+
+    class SubscriptFinder(IdentityMapper):
+
+        def __init__(self, find_aggregate_name, find_index_names):
+            self.find_aggregate_aname = find_aggregate_name
+            self.find_index_names = find_index_names
+            self.result = False
+
+        def map_subscript(self, expr, *args, **kwargs):
+            if expr.aggregate.name == self.find_aggregate_aname:
+                vf = VariableFinder(self.find_index_names)
+                vf(expr.index)
+                if vf.result:
+                    self.result = True
+            return super(SubscriptFinder, self).map_subscript(expr, args, kwargs)
+
+    simd_inames = []
+    new_simd_inames = []
+
+    for i in sorted(kernel.all_inames()):
+        if kernel.iname_tags_of_type(i, OpenMPSIMDTag):
+            simd_inames.append(i)
+            j = i + "_p"
+            new_simd_inames.append(j)
+
+            # create domains for new inames
+            domch = DomainChanger(kernel, frozenset([i]))
+            kernel = kernel.copy(domains=domch.get_domains_with(
+                duplicate_axes(domch.domain, [i], [j])))
+
+            # change omp_simd to ilp.seq for non-vectorizable instructions
+            kernel = tag_inames(kernel, [(i, "ilp.seq")], retag=True)
+            kernel = tag_inames(kernel, [(j, "omp_simd")])
+
+    if not simd_inames:
+        return kernel
+
+    iname_map = dict(zip(simd_inames, new_simd_inames))
+    subst_mapper = SubstitutionMapper(
+        make_subst_func(dict((Variable(o), Variable(n))
+                             for (o, n) in zip(simd_inames, new_simd_inames))))
+    globals = [name for name, arg in kernel.arg_dict.items()
+               if isinstance(arg, ArrayArg) and arg.shape[0] is not None]
+    gbf = VariableFinder(globals)
+
+    new_insts = []
+    for inst in kernel.instructions:
+        if isinstance(inst, Assignment) and (inst.within_inames & set(simd_inames)):
+            should_simd = True
+
+            # reduction over globals are sequantialized
+            if should_simd and globals:
+                gbf.result = False
+                gbf(inst.assignee)
+                if gbf.result:
+                    should_simd = False
+
+            # cannot have inames outside of subscript
+            if should_simd:
+                ovf = OutsideVariableFinder(
+                    list(inst.within_inames & set(simd_inames)))
+                ovf(inst.expression)
+                if ovf.result:
+                    should_simd = False
+
+            # quick hack to turn off simd for gathering
+            # this is OK since omp vectorization will note be used in practice
+            # this is just for gathering data for the paper
+            if inst.assignee.aggregate.name[:3] == "dat":
+                should_simd = False
```

Comment:
This should be something like "checking if the LHS is stride one", in particular, here we need to get ride of the `A[B[i]]` case.

Reply:
Also @kaushikcfd 's labels in TSFC might be useful here.